### DISCUSSION TITLE
U can setup base middleware system smth like this:

### DISCUSSION THREAD

**Discussion by haoynmail** [uNetworking/uWebSockets.js#1050] [Q&A]
U can setup base middleware system smth like this:

```javascript
const middlewares = [
   (res, req, next) => { ... },
   (res, req, next) => { ... }
];

const runMiddlewares = (res, req) => {
    let index = 0;

    const next = async () => {
        const nextMiddleware = middlewares[index++];
        const hasRespondEnded = res.finished; // this is CUSTOM property, see my comment below
        const shouldRunNextMiddleware = !hasRespondEnded && nextMiddleware;

        if (shouldRunNextMiddleware) {
            nextMiddleware(res, req, next);
        }
    };

    next();
};

App = new App().any('/*', runMiddlewares)

```
Unfortunately, today's HttpResponse API doesnt provide easy way to check if response has finished.
For now you propably might want to override the res.end and res.tryEnd methods:

```javascript
const enhanceRes = (res) => {
     const originalEnd = res.end.bind(res);
     const originalTryEnd = res.tryEnd.bind(res);
     
     res.finished = false;
     
     res.end = (...args) => {
         originalEnd(...args);
         res.finished = true;
         return res;
     }
     
     res.tryEnd = (...args) => {
          const [ok, done] = originalTryEnd(...args);
          res.finished = done;
          return [ok, done]
     }
}
```

_Originally posted by @kolodziejczak-sz in https://github.com/uNetworking/uWebSockets.js/discussions/483#discussioncomment-533775_

**Comment by haoynmail** [comment]
hello @kolodziejczak-sz

Please advise on the use about  **_const enhanceRes=(res)=>{...}_**.

For example, I have a middleware array:
 const middleware=[
    (res, req, next)=>{...},
   (res, req, next)=>{...}, (this position res.end (...), not executed later)
   (res, req, next)=>{...},
];

I want the second middleware to send messages to the browser, while the third middleware does not execute. How to use **_enhancRes (res)_**

**Comment by DadiBit** [comment]
Also, please note that you can/should use a Proxy on apply. It automatically infers all types in typescript too!

**Reply by haoynmail** [reply]
Thank you for your reply.

I didn't understand your meaning. I don't want to do a type check.

I mean: how does the function enhanceRes=(res)=>{...}  execute in the middleware system mentioned above.
For example: there are three routing functions in the middleware system

const middleware=[
  (res, req, next)=>{...},
  (res, req, next)=>{...}
  (res, req, next)=>{...},
];
How do use  the function enhanceRes=(res)=>{...}  to make routing functions 1, 2 to run ,but 3 not.  Or how do use  the function enhanceRes=(res)=>{...} to make routing functions 1, 3 to run, but 2 not.
